File: Java/chapter_12_challenge/app/src/main/java/com/glella/criminalintent/TimePickerFragment.java
package com.glella.criminalintent;

import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TimePicker;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

// Challenge - This class and dialog_time layout plus modification of fragment_crime layout to have 2 buttons
// one for the date and another for the time

public class TimePickerFragment extends DialogFragment {

    private static final String ARG_TIME = "time";

    public static final String EXTRA_TIME = "com.glella.criminalintent.time";

    private TimePicker mTimePicker;

    public static TimePickerFragment newInstance(Date date) {
        Bundle args = new Bundle();
        args.putSerializable(ARG_TIME, date);

        TimePickerFragment fragment = new TimePickerFragment();
        fragment.setArguments(args);
        return fragment;
    }


    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {

        Date date = (Date) getArguments().getSerializable(ARG_TIME);
        final Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);

        int hours = calendar.get(Calendar.HOUR_OF_DAY);
        int minutes = calendar.get(Calendar.MINUTE);

        View v = LayoutInflater.from(getActivity()).inflate(R.layout.dialog_time, null);

        mTimePicker = (TimePicker) v.findViewById(R.id.dialog_time_picker);
        mTimePicker.setHour(hours);
        mTimePicker.setMinute(minutes);

        return new AlertDialog.Builder(getActivity())
                .setView(v)
                .setTitle(R.string.time_picker_title)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int which) {
                        int year = calendar.get(Calendar.YEAR);
                        int month = calendar.get(Calendar.MONTH);
                        int day = calendar.get(Calendar.DAY_OF_MONTH);
                        int hours = mTimePicker.getHour();
                        int minutes = mTimePicker.getMinute();
                        Date date = new GregorianCalendar(year, month, day, hours, minutes).getTime();

                        sendResult(Activity.RESULT_OK, date);
                    }
                })
                .create();
    }

    private void sendResult(int resultCode, Date date) {
        if (getTargetFragment() == null) {
            return;
        }

        Intent intent = new Intent();
        intent.putExtra(EXTRA_TIME, date);

        getTargetFragment().onActivityResult(getTargetRequestCode(), resultCode, intent);
    }
}

File: Kotlin/chapter_12_challenge/app/src/main/java/com/glella/criminalintent/TimePickerFragment.kt
package com.glella.criminalintent

import android.app.Activity
import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v7.app.AlertDialog
import android.view.LayoutInflater
import android.view.View
import android.widget.TimePicker
import java.util.*

// Challenge - This class and dialog_time layout plus modification of fragment_crime layout to have 2 buttons
// one for the date and another for the time

class TimePickerFragment : DialogFragment() {

    companion object {

        private val ARG_TIME = "time"

        val EXTRA_TIME = "com.glella.criminalintent.time"

        fun newInstance(date: Date): TimePickerFragment {
            val args = Bundle()
            args.putSerializable(ARG_TIME, date)

            val fragment = TimePickerFragment()
            fragment.arguments = args
            return fragment
        }
    }

    lateinit var mTimePicker: TimePicker

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {

        val date = arguments!!.getSerializable(ARG_TIME) as Date
        val calendar = Calendar.getInstance()
        calendar.time = date

        val hours = calendar.get(Calendar.HOUR_OF_DAY)
        val minutes = calendar.get(Calendar.MINUTE)

        val v = LayoutInflater.from(activity).inflate(R.layout.dialog_time, null)

        mTimePicker = v.findViewById<View>(R.id.dialog_time_picker) as TimePicker
        mTimePicker.hour = hours
        mTimePicker.minute = minutes

        return AlertDialog.Builder(activity!!)
                .setView(v)
                .setTitle(R.string.time_picker_title)
                .setPositiveButton(android.R.string.ok) { dialogInterface, which ->
                    val year = calendar.get(Calendar.YEAR)
                    val month = calendar.get(Calendar.MONTH)
                    val day = calendar.get(Calendar.DAY_OF_MONTH)
                    val hours = mTimePicker.hour
                    val minutes = mTimePicker.minute
                    val date = GregorianCalendar(year, month, day, hours, minutes).time

                    sendResult(Activity.RESULT_OK, date)
                }
                .create()

    }

    private fun sendResult(resultCode: Int, date: Date) {
        if (targetFragment == null) {
            return
        }

        val intent = Intent()
        intent.putExtra(EXTRA_TIME, date)

        targetFragment!!.onActivityResult(targetRequestCode, resultCode, intent)
    }

}